flink-runtime/src/main/java/org/apache/flink/runtime/memory/MemoryManager.java

```
@@ -474,67 +424,141 @@ public void releaseAll(Object owner) {
+	 * @param size size of memory to reserve.
+	 * @throws MemoryAllocationException Thrown, if this memory manager does not have the requested amount
+	 *                                   of memory any more.
+	 */
+	public void reserveMemory(Object owner, MemoryType memoryType, long size) throws MemoryAllocationException {
+		checkMemoryReservationPreconditions(owner, memoryType, size);
+		if (size == 0L) {
+			return;
+		}
+
+		long acquiredMemory = budgetByType.acquireBudgetForKey(memoryType, size);
+		if (acquiredMemory < size) {
+			throw new MemoryAllocationException(
+				String.format("Could not allocate %d bytes. Only %d bytes are remaining.", size, acquiredMemory));
+		}
+
+		reservedMemory.compute(owner, (o, reservations) -> {
+			Map<MemoryType, Long> newReservations = reservations;
+			if (reservations == null) {
+				newReservations = new EnumMap<>(MemoryType.class);
+				newReservations.put(memoryType, size);
+			} else {
+				reservations.compute(
+					memoryType,
+					(mt, currentlyReserved) -> currentlyReserved == null ? size : currentlyReserved + size);
+			}
+			return newReservations;
+		});
+
+		Preconditions.checkState(!isShutDown, "Memory manager has been concurrently shut down.");
+	}
 
 	/**
-	 * Gets the type of memory (heap / off-heap) managed by this memory manager.
+	 * Releases memory of a certain type from an owner to this memory manager.
 	 *
-	 * @return The type of memory managed by this memory manager.
+	 * @param owner The owner to associate with the memory reservation, for the fallback release.
+	 * @param memoryType type of memory to release (heap / off-heap).
+	 * @param size size of memory to release.
 	 */
-	public MemoryType getMemoryType() {
-		return memoryType;
+	public void releaseMemory(Object owner, MemoryType memoryType, long size) {
+		checkMemoryReservationPreconditions(owner, memoryType, size);
+		if (size == 0L) {
+			return;
+		}
+
+		reservedMemory.compute(owner, (o, reservations) -> {
```

It seems we're using a "fail silently" way here if the `compute` result is `null` which means we cannot find the owner in the map anymore. Is it possible that one releases its reserved memory multiple times? If so, does it worth any logging here?

---

As I understand, we already have a log message for this:
```
LOG.warn(
	"Trying to release more memory {} than it was reserved {} so far for the owner {}",
	size,
	currentlyReserved == null ? 0 : currentlyReserved,
	owner);
```
or do you mean something else?
or the older code was reviewed :)